node/src/components/network/conman.rs

```diff
@@ -0,0 +1,1054 @@
+    /// `ConMan` is dropped.
+    shutdown: DropSwitch<ObservableFuse>,
+}
+
+#[derive(Copy, Clone, Debug)]
+/// Configuration settings for the connection manager.
+struct Config {
+    /// The timeout for one TCP to be connection to be established, from a single `connect` call.
+    tcp_connect_timeout: Duration,
+    /// Maximum time allowed for TLS setup and handshaking to proceed.
+    setup_timeout: Duration,
+    /// How often to reattempt a connection.
+    ///
+    /// At one second, 8 attempts means that the last attempt will be delayed for 128 seconds.
+    tcp_connect_attempts: NonZeroUsize,
+    /// Base delay for the backoff, grows exponentially until `tcp_connect_attempts` maxes out.
+    tcp_connect_base_backoff: Duration,
+    /// How long to back off from reconnecting to an address after a failure that indicates a
+    /// significant problem.
+    significant_error_backoff: Duration,
+    /// How long to back off from reconnecting to an address if the error is likely not going to
+    /// change for a long time.
+    permanent_error_backoff: Duration,
+    /// How long to wait before reconnecting when a succesful outgoing connection is lost.
+    successful_reconnect_delay: Duration,
+    /// The minimum time a connection must have successfully served data to not be seen as flaky.
+    flaky_connection_threshold: Duration,
+    /// Number of incoming connections before refusing to accept any new ones.
+    max_incoming_connections: usize,
+    /// Number of outgoing connections before stopping to connect to learned addresses.
+    max_outgoing_connections: usize,
+}
+
+/// Shared information across the connection manager and its subtasks.
+struct ConManContext {
+    /// Shared configuration settings.
+    cfg: Config,
+    /// Callback handler for connection setup and incoming request handling.
+    protocol_handler: Box<dyn ProtocolHandler>,
+    /// Juliet RPC configuration.
+    rpc_builder: RpcBuilder<{ super::Channel::COUNT }>,
+    /// The shared state.
+    state: RwLock<ConManState>,
+    /// Our own address (for loopback filtering).
+    public_addr: SocketAddr,
+    /// Our own node ID.
+    our_id: NodeId,
+    /// Limiter for incoming connections.
+    incoming_limiter: Arc<Semaphore>,
+}
+
+/// Share state for [`ConMan`].
+///
+/// Tracks outgoing and incoming connections.
+#[derive(Debug, Default)]
+pub(crate) struct ConManState {
+    /// A set of outgoing address for which a handler is currently running.
+    address_book: HashSet<SocketAddr>,
+    /// Mapping of [`SocketAddr`]s to an instant in the future until which they must not be dialed.
+    do_not_call: HashMap<SocketAddr, Instant>,
+    /// The current route per node ID.
+    ///
+    /// An entry in this table indicates an established connection to a peer. Every entry in this
+    /// table is controlled by an `OutgoingHandler`, all other access should be read-only.
+    routing_table: HashMap<NodeId, Route>,
```

Comment:
That overloads the term "connection" a little too much, since a `Route` is not just a connection. I thought of reaching a certain node (by node id) _via_ a route (which is IP based).

To be honest, I think it is similar to [ARP](https://en.wikipedia.org/wiki/Address_Resolution_Protocol), if one considers the node address unchanging, but the connection (not address!) changing instead.

Frankly, I don't have a better suggestion at the moment, except that I don't think `connection_table` is better.